### PR TITLE
MINIFICPP-2341 Include custom relationships in C2 manifest for NiFi python processors

### PR DESCRIPTION
**Pull request by lordgamez**
- Additionally adapt some python classes more closely to NiFi's API

https://issues.apache.org/jira/browse/MINIFICPP-2341

----------------------------
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: extensions/python/pythonprocessors/nifiapi/properties.py [hunk 1]
```diff
@@ -281,3 +284,21 @@ def getProperty(self, descriptor) -> PythonPropertyValue:
...
+
+    def getProperties(self) -> Dict[PropertyDescriptor, str]:
+        properties = dict()
+        cpp_properties = self.cpp_context.getProperties()
+
+        for property_name in cpp_properties:
+            for property_descriptor in self.processor.getPropertyDescriptors():
+                if property_descriptor.name == property_name:
+                    properties[property_descriptor] = cpp_properties[property_name]
+                    break
```
Comment (szaszm) [inline extensions/python/pythonprocessors/nifiapi/properties.py:302]: This is quadratic, we should improve this later to have linear complexity. It's fine for now, but I made a Jira issue to fix this later: [MINIFICPP-2361](https://issues.apache.org/jira/browse/MINIFICPP-2361)

**Review by szaszm — APPROVED** [review]

File: extensions/python/PYTHON.md [hunk 2]
```diff
@@ -166,6 +166,10 @@ Due to some differences between the NiFi and MiNiFi C++ processors and implement
 - MiNiFi C++ only supports expression language with flow file attributes, so only FLOWFILE_ATTRIBUTES expression language scope is supported, otherwise the expression language will not be evaluated.
 - MiNiFi C++ does not support property dependencies, so the property dependencies will be ignored. If a property depends on another property, the property will not be required.
 - MiNiFi C++ does not support the use of self.jvm member in Python processors that provides JVM bindings in NiFi, it is set to None in MiNiFi C++.
+- Dynamic properties are supported in all Python processors and the dynamic properties defined in the flow configuration are automatically added. There is no need to define getDynamicPropertyDescriptor method in the Python processor. The only caveat is that the description of the dynamic property cannot be custom defined.
+- In MiNiFi C++ when the processor is stopped the stop event handling is done in the `notifystop()` processor which does not have the context available. Due to this the `def onStopped(self, context)` cannot be called in NiFi Python processors, so the `onStopped` method is not supported in MiNiFi C++.
```
Comment (adamdebreceni) [inline extensions/python/PYTHON.md:170]: do you mean the following?
```suggestion
- In MiNiFi C++ when the processor is stopped the stop event handling is done in the `notifyStop()` method which does not have the context available. Due to this the `def onStopped(self, context)` cannot be called in NiFi Python processors, so the `onStopped` method is not supported in MiNiFi C++.
```

File: extensions/python/pythonprocessors/nifiapi/properties.py (hunk 1 above)
Comment (adamdebreceni) [reply extensions/python/pythonprocessors/nifiapi/properties.py:310]: can't we just flip it and do, `cpp_properties` seems to be a wrapped python dictionary, forwarding the `contains` call (which I think the following code would do)
```suggestion
        for property_descriptor in self.processor.getPropertyDescriptors():
            if property_descriptor.name in cpp_properties:
                properties[property_descriptor] = cpp_properties[property_name]
```

File: extensions/python/PYTHON.md (hunk 2 above)
Comment (lordgamez) [reply extensions/python/PYTHON.md]: Yes, good catch, updated in 6015abfc8d282d8e9f9ef7458de4b4686298d696

File: extensions/python/pythonprocessors/nifiapi/properties.py (hunk 1 above)
Comment (lordgamez) [reply extensions/python/pythonprocessors/nifiapi/properties.py]: I updated the test and the implementation of getProperties in 9eec8f9b0a61264578330919ed9e16fa2c5007fd

Comment (szaszm) [reply extensions/python/pythonprocessors/nifiapi/properties.py]: That's a hash table lookup now, right? That would be amortized linear complexity, which is fine IMO. If you can confirm, I'll close the Jira issue.

Comment (adamdebreceni) [reply extensions/python/pythonprocessors/nifiapi/properties.py]: yes, I think the Jira issue can be closed now

**Review by adamdebreceni — APPROVED** [review]

File: docker/test/integration/minifi/processors/ProcessContextInterfaceChecker.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+
+class ProcessContextInterfaceChecker(Processor):
+    def __init__(self, context):
+        super(ProcessContextInterfaceChecker, self).__init__(
+            context=context,
+            clazz='ProcessContextInterfaceChecker',
+            class_prefix='org.apache.nifi.minifi.processors.nifi_python_processors.',
+            properties={},
+            auto_terminate=[])
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply docker/test/integration/minifi/processors/ProcessContextInterfaceChecker.py]: Good point, updated in 195fbb63b74349665337f146ad33bbf10bfd061e

**Review by fgerlits — APPROVED** [review]

**Review by martinzink — APPROVED** [review]

**Pull request closed without merging**